motion_primitives_forward_controller/src/motion_primitives_forward_controller.cpp

```diff
@@ -0,0 +1,554 @@
+    std::this_thread::sleep_for(std::chrono::milliseconds(10));
+  }
+
+  reset_command_interfaces();
+  // send reset stop command to the hw-interface
+  (void)command_interfaces_[0].set_value(static_cast<double>(MotionType::RESET_STOP));
+  robot_stopped_ = false;
+
+  RCLCPP_INFO(get_node()->get_logger(), "Robot stopped, ready for new motion primitives.");
+  return rclcpp_action::CancelResponse::ACCEPT;
+}
+
+void MotionPrimitivesForwardController::goal_accepted_callback(
+  const std::shared_ptr<rclcpp_action::ServerGoalHandle<ExecuteMotion>> goal_handle)
+{
+  std::thread{std::bind(&MotionPrimitivesForwardController::execute_goal, this, goal_handle)}
+    .detach();
+}
+
+void MotionPrimitivesForwardController::execute_goal(
+  const std::shared_ptr<rclcpp_action::ServerGoalHandle<ExecuteMotion>> goal_handle)
+{
+  const auto & goal = goal_handle->get_goal();
+
+  if (robot_stopped_)
+  {
+    RCLCPP_WARN(get_node()->get_logger(), "Robot is stopped. Discarding the new command.");
+    return;
+  }
+
+  const auto & primitives = goal->trajectory.motions;
+
+  if (primitives.empty())
+  {
+    RCLCPP_WARN(get_node()->get_logger(), "Received goal with no motion primitives. Ignoring.");
+    return;
+  }
+
+  size_t required_queue_space = primitives.size();
+  // add sequence markers only if there are multiple primitives
+  bool add_sequence_wrappers = primitives.size() > 1;
+  if (add_sequence_wrappers)
+  {
+    required_queue_space += 2;  // start + end
+  }
+
+  // Check if queue has enough space
+  if (moprim_queue_.size() + required_queue_space > queue_size_)
+  {
+    RCLCPP_ERROR(
+      get_node()->get_logger(),
+      "Received more motion primitives than queue can store. "
+      "Current queue size: %zu, required space: %zu. "
+      "Please increase the queue size in the controller parameters.",
+      moprim_queue_.size(), required_queue_space);
+    return;
+  }
+
+  // Add sequence start marker
+  if (add_sequence_wrappers)
+  {
+    std::shared_ptr<MotionPrimitiveMsg> start_marker = std::make_shared<MotionPrimitiveMsg>();
+    start_marker->type = MotionType::MOTION_SEQUENCE_START;
+    moprim_queue_.push(start_marker);
+  }
+
+  // Add motion primitives to the queue
+  for (const auto & primitive : primitives)
+  {
+    moprim_queue_.push(std::make_shared<MotionPrimitiveMsg>(primitive));
+  }
+
+  // Add sequence end marker
+  if (add_sequence_wrappers)
+  {
+    std::shared_ptr<MotionPrimitiveMsg> end_marker = std::make_shared<MotionPrimitiveMsg>();
+    end_marker->type = MotionType::MOTION_SEQUENCE_END;
+    moprim_queue_.push(end_marker);
+  }
+
+  RCLCPP_INFO(
+    get_node()->get_logger(), "Accepted goal with %zu motion primitives%s.", primitives.size(),
+    add_sequence_wrappers ? " (wrapped in sequence markers)" : "");
+
+  ExecuteMotion::Result result;
+  rclcpp::Rate rate(50);
+  while (rclcpp::ok())
+  {
+    auto execution_status = static_cast<uint8_t>(std::round(state_interfaces_[0].get_value()));
```

Comment:
You are basically accessing a shared resource (`state_interfaces_` gets set inside the control thread) without thread synchronization. One more point against using the execute thread.